opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/internal/tracer/MetricsTracer.java

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright 2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.contrib.metrics.internal.tracer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.opentracing.ActiveSpan;
+import io.opentracing.BaseSpan;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.contrib.metrics.MetricsReporter;
+import io.opentracing.propagation.Format;
+
+public class MetricsTracer implements Tracer {
+
+    private final Tracer tracer;
+    private final MetricsReporter reporter;
+
+    private final Map<SpanContext,MetricData> metricData = new WeakHashMap<SpanContext,MetricData>();
```

Comment:
The wrapped tracer maybe dealing with multiple instances of ActiveSpan that map onto a single actual span context - so the SpanContext is the only reliable key.